cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -620,6 +620,30 @@ func (c *machineController) findScalableResourceProviderIDs(scalableResource *un
 	}
 
 	for _, machine := range machines {
+		// In some cases it is possible for a machine to have acquired a provider ID from the infrastructure and
+		// then become failed later. We want to ensure that a failed machine is not counted towards the total
+		// number of nodes in the cluster, for this reason we will detect a failed machine first, regardless
+		// of provider ID, and give it a normalized provider ID with failure message prepended.
+		failureMessage, found, err := unstructured.NestedString(machine.UnstructuredContent(), "status", "failureMessage")
+		if err != nil {
+			return nil, err
+		}
+
+		if found {
+			klog.V(4).Infof("Status.FailureMessage of machine %q is %q", machine.GetName(), failureMessage)
+			// Provide a fake ID to allow the autoscaler to track machines that will never
+			// become nodes and mark the nodegroup unhealthy after maxNodeProvisionTime.
+			// Fake ID needs to be recognised later and converted into a machine key.
+			// Use an underscore as a separator between namespace and name as it is not a
+			// valid character within a namespace name.
+			providerIDs = append(providerIDs, fmt.Sprintf("%s%s_%s", failedMachinePrefix, machine.GetNamespace(), machine.GetName()))
```

Comment:
should we be able to use a common string concatenation approach here as in the earlier PR? in which case we'd have one func that is easily micro-unit testable and not do any inline `fmt.Sprintf` foo whose result is more or less part of the machine data contract.

Reply:
we can, but then i feel we should refactor all the uses of this style of prepending. there are a few other places in the code that could be updated as well.

Reply:
One more question: why aren't we able to add another condition to L656 below:

`if !machine.GetDeletionTimestamp().IsZero() && !failureMessage` (for example)

That would short-circuit that machine and exclude it from the providerIDs response slice. Would that produce the desired outcome?

Reply:
added the function and test, it turns out this was not used as widely as i had feared.

Reply:
apologies, i resolved this and hadn't seen the new message.

> That would short-circuit that machine and exclude it from the providerIDs response slice. Would that produce the desired outcome?

we would still need to add the clause for detecting a failed machine later, after the "no provider id" warning. it seemed to me that we always want to short circuit when a machine is failed. we also need to differentiate between a deleting machine and failed machine.